src/utils/format-numbers.js

```diff
@@ -237,92 +237,143 @@ export function formatMilliseconds(
+    result = timeInSeconds.toFixed(digits);
+  } else {
+    result = formatNumber(
+      timeInSeconds,
+      significantDigits,
+      maxFractionalDigits
+    );
+  }
+  return result + 's';
 }
 
 export function formatMinutes(
   time: Milliseconds,
   significantDigits: number = 5,
-  maxFractionalDigits: number = 2
+  maxFractionalDigits: number = 2,
+  precision: Milliseconds = Infinity
 ) {
   const msPerSecond = 1000;
-  time = Math.round(time / msPerSecond) * msPerSecond;
   const msPerMinute = 60 * msPerSecond;
+  if (precision >= msPerSecond) {
+    time = Math.round(time / msPerSecond) * msPerSecond;
+  }
   const seconds = time % msPerMinute;
   return (
     formatNumber((time - seconds) / msPerMinute, significantDigits, 0) +
     'min' +
-    (maxFractionalDigits > 0 && seconds > 0
-      ? formatSeconds(seconds, significantDigits, 0)
+    (seconds > 0 && (maxFractionalDigits > 0 || precision < msPerMinute)
+      ? formatSeconds(seconds, significantDigits, 0, precision)
       : '')
   );
 }
 
 export function formatHours(
   time: Milliseconds,
   significantDigits: number = 5,
-  maxFractionalDigits: number = 1
+  maxFractionalDigits: number = 1,
+  precision: Milliseconds = Infinity
 ) {
   const msPerMinute = 60 * 1000;
-  time = Math.round(time / msPerMinute) * msPerMinute;
+  if (precision >= msPerMinute) {
+    time = Math.round(time / msPerMinute) * msPerMinute;
+  }
   const msPerHour = 60 * msPerMinute;
   const minutes = time % msPerHour;
   return (
     formatNumber((time - minutes) / msPerHour, significantDigits, 0) +
     'h' +
-    (maxFractionalDigits > 0 && minutes > 0
-      ? formatMinutes(minutes, significantDigits, 0)
+    (minutes > 0 && (maxFractionalDigits > 0 || precision < msPerHour)
+      ? formatMinutes(minutes, significantDigits, 0, precision)
       : '')
   );
 }
 
 export function formatDays(
   time: Milliseconds,
   significantDigits: number = 5,
-  maxFractionalDigits: number = 1
+  maxFractionalDigits: number = 1,
+  precision: Milliseconds = Infinity
 ) {
   const msPerHour = 60 * 60 * 1000;
-  time = Math.round(time / msPerHour) * msPerHour;
+  if (precision >= msPerHour) {
+    time = Math.round(time / msPerHour) * msPerHour;
+  }
   const msPerDay = 24 * msPerHour;
   const hours = time % msPerDay;
   return (
     formatNumber((time - hours) / msPerDay, significantDigits, 0) +
     'd' +
-    (maxFractionalDigits > 0 && hours > 0
-      ? formatHours(hours, significantDigits, 0)
+    (hours > 0 && (maxFractionalDigits > 0 || precision < msPerDay)
+      ? formatHours(hours, significantDigits, 0, precision)
       : '')
   );
 }
 
 export function formatTimestamp(
   time: Milliseconds,
   significantDigits: number = 5,
-  maxFractionalDigits: number = 3
+  maxFractionalDigits: number = 3,
+  // precision is the minimum required precision.
+  precision: Milliseconds = Infinity
 ) {
+  if (precision !== Infinity) {
+    precision = 10 ** Math.floor(Math.log10(precision));
```

Comment:
I just wonder if this part should be done in the caller, so that the `precision` arguments have the same meaning in all the functions here. What do you think?

Although I'm not 100% sure about this line does. Would we have too much precision without this line?

Reply:
I think doing it in the caller would be more effort once there are multiple callers (I'm hoping the `TimelineRuler` code will eventually use this code path too). The `precision` argument still works in the same way in all the functions, thanks to the `Math.floor(Math.log10(` call in `formatSeconds`.

I added a comment in the code to explain what behavior these lines implement. Not the most useful part of the patch, unless it was needed to fix an edge case I don't remember.